### PR TITLE
Add video keyframe extraction

### PR DESCRIPTION
**Pull request by wonjuleee**
### Summary

Depends on [#100408](https://jira.devtools.intel.com/browse/CVS-100408).
- Added ZNCC-based keyframe extractor feature

### How to test

datum create -o proj
datum import -f video_keyframes -p proj video.mp4 -t 0.3 --
datum export -p proj -f image_dir --

### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md)
- [ ] I have updated the [documentation](
  https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly
- [x] I have added tests to cover my changes
- [ ] I have [linked related issues](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: datumaro/plugins/data_formats/video.py [hunk 1]
```diff
@@ -124,3 +127,87 @@ def __iter__(self):
...
+        return parser
+
+    @classmethod
+    def find_sources(cls, path):
+        if not osp.isfile(path):
+            return []
+        return [{"url": path, "format": VideoKeyframesBase.NAME}]
+
+
+_FEAT_SIZE = (64, 64)
```
Comment (vinnamkim) [inline datumaro/plugins/data_formats/video.py]: Can it be configurable?

File: datumaro/plugins/data_formats/video.py [hunk 2]
```diff
@@ -124,3 +127,87 @@ def __iter__(self):
...
+            return True
+
+        _curr_frame = cv2.resize(cv2.cvtColor(np.uint8(frame), cv2.COLOR_BGR2GRAY), _FEAT_SIZE)
+        _curr_stats = cv2.meanStdDev(_curr_frame)
+
+        zncc_score = np.sum(
+            np.multiply((self._keyframe - self._keyframe_stats[0]), (_curr_frame - _curr_stats[0])),
+            dtype=np.float32,
+        )
+        zncc_score /= self._keyframe_stats[1] * _curr_stats[1] * _FEAT_SIZE[0] * _FEAT_SIZE[1]
```
Comment (vinnamkim) [inline datumaro/plugins/data_formats/video.py]: We can add a small epsilon, e.g., 1e-6 for numerical stability for division.

Comment (wonjuleee) [reply datumaro/plugins/data_formats/video.py]: thanks. I added the epsilon as you suggested.

File: datumaro/plugins/data_formats/video.py (hunk 1 above)
Comment (wonjuleee) [reply datumaro/plugins/data_formats/video.py]: I added a new parameter for adjusting the image size during the ZNCC computation.

File: datumaro/plugins/data_formats/video.py [hunk 3]
```diff
@@ -138,7 +138,23 @@ class VideoKeyframesImporter(VideoFramesImporter):
...
+            "--threshold",
+            type=float,
+            default=0.3,
+            help="Similarity threshold (default: %(default)s)",
+        )
+        return parser
+
+    @classmethod
+    def build_cmdline_parser(cls, **kwargs):
+        parser = super().build_cmdline_parser(**kwargs)
```
Comment (vinnamkim) [inline datumaro/plugins/data_formats/video.py]:
```suggestion
```

It seems that those lines should be removed not to declare `build_cmdline_parser()` again.

**Review by bonhunko — DISMISSED** [review]
LGTM!

**Review by sooahleex — DISMISSED** [review]
Looks good to me. Just one minor comment, I think It seems good to add comment for adding epsilon.

File: datumaro/plugins/data_formats/video.py (hunk 3 above)
Comment (wonjuleee) [reply datumaro/plugins/data_formats/video.py]: thank you for pointing out this!

File: tests/cli/test_video.py [hunk 4]
```diff
@@ -50,3 +51,74 @@ def test_can_display_video_import_warning_in_add(self):
...
+                "2",
+            )
+
+            result_dir = osp.join(proj_dir, "result")
+            run(self, "export", "-f", "image_dir", "-p", proj_dir, "-o", result_dir)
+            parsed_dataset = Dataset.import_from(result_dir, "image_dir")
+
+            for d in parsed_dataset:
+                print(d)
+
```
Comment (vinnamkim) [inline tests/cli/test_video.py]:
```suggestion
```
We can remove these lines.

File: datumaro/plugins/data_formats/video.py [hunk 5]
```diff
@@ -124,3 +127,101 @@ def __iter__(self):
...
+        parser.add_argument(
+            "-t",
+            "--threshold",
+            type=float,
+            default=0.3,
+            help="Similarity threshold (default: %(default)s)",
+        )
+
+        parser.add_argument(
+            "-resize",
```
Comment (vinnamkim) [inline datumaro/plugins/data_formats/video.py]:
```suggestion
            "-r",
```

It's common practice to combine a single hyphen with a single character. Please refer to https://serverfault.com/a/387936/776078.

File: datumaro/plugins/data_formats/video.py [hunk 6]
```diff
@@ -124,3 +127,101 @@ def __iter__(self):
...
+        self._subset = subset or DEFAULT_SUBSET_NAME
+        super().__init__(
+            url=url,
+            step=step,
+            start_frame=start_frame,
+            end_frame=end_frame,
+            name_pattern=name_pattern,
+        )
+
+        assert osp.isfile(url), url
```
Comment (vinnamkim) [inline datumaro/plugins/data_formats/video.py]:
```suggestion
        super().__init__(
            url=url,
            subset=subset,
            step=step,
            start_frame=start_frame,
            end_frame=end_frame,
            name_pattern=name_pattern,
        )
```

Some lines are redundant in `VideoFramesBase.__init__()`

File: datumaro/plugins/data_formats/video.py [hunk 7]
```diff
@@ -124,3 +127,101 @@ def __iter__(self):
...
+            self._keyframe = cv2.resize(
+                cv2.cvtColor(np.uint8(frame), cv2.COLOR_BGR2GRAY), self._resize
+            )
+            self._keyframe_stats = cv2.meanStdDev(self._keyframe)
+            return True
+
+        _curr_frame = cv2.resize(cv2.cvtColor(np.uint8(frame), cv2.COLOR_BGR2GRAY), self._resize)
+        _curr_stats = cv2.meanStdDev(_curr_frame)
+
+        zncc_score = np.sum(
```
Comment (vinnamkim) [inline datumaro/plugins/data_formats/video.py]: We can use `np.mean()` and remove division by `self._resize[0] * self._resize[1]` below.

File: datumaro/plugins/data_formats/video.py [hunk 8]
```diff
@@ -124,3 +127,101 @@ def __iter__(self):
...
+            np.multiply((self._keyframe - self._keyframe_stats[0]), (_curr_frame - _curr_stats[0])),
```
Comment (vinnamkim) [inline datumaro/plugins/data_formats/video.py]: We can store `self._keyframe - self._keyframe_stats[0]` whenever the keyframe is updated and reuse it.

File: datumaro/plugins/data_formats/video.py [hunk 9]
```diff
@@ -124,3 +127,101 @@ def __iter__(self):
...
+        assert osp.isfile(url), url
+
+        self._threshold = threshold
+        self._keyframe = None
+        self._keyframe_stats = None
+        self._resize = resize
+
+    def _is_keyframe(self, frame):
+        if self._keyframe is None:
+            self._keyframe = frame
```
Comment (vinnamkim) [inline datumaro/plugins/data_formats/video.py]: We can remove this line.

File: datumaro/plugins/data_formats/video.py (hunk 5 above)
Comment (wonjuleee) [reply datumaro/plugins/data_formats/video.py]: thank you, I corrected

File: datumaro/plugins/data_formats/video.py (hunk 6 above)
Comment (wonjuleee) [reply datumaro/plugins/data_formats/video.py]: thank you, I corrected.

File: datumaro/plugins/data_formats/video.py (hunk 7 above)
Comment (wonjuleee) [reply datumaro/plugins/data_formats/video.py]: I just got identified some mal-implementations. I optimized the code.

File: datumaro/plugins/data_formats/video.py (hunk 8 above)
Comment (wonjuleee) [reply datumaro/plugins/data_formats/video.py]: here too. Thank you.

**Review by vinnamkim — APPROVED** [review]
LGTM.

**Pull request merged**